File: solutions/fable-1/fable/perl/calc.pl
use strict;
use warnings;

my @tok;
my $pos;

sub parse_expr {
    my $v = parse_term();
    while ($pos < @tok && ($tok[$pos] eq '+' || $tok[$pos] eq '-')) {
        my $op = $tok[$pos++];
        my $r  = parse_term();
        $v = $op eq '+' ? $v + $r : $v - $r;
    }
    return $v;
}

sub parse_term {
    my $v = parse_factor();
    while ($pos < @tok && ($tok[$pos] eq '*' || $tok[$pos] eq '/')) {
        my $op = $tok[$pos++];
        my $r  = parse_factor();
        if ($op eq '*') {
            $v = $v * $r;
        } else {
            my $q = int(abs($v) / abs($r));
            $q = -$q if ($v < 0) != ($r < 0);
            $v = $q;
        }
    }
    return $v;
}

sub parse_factor {
    my $t = $tok[$pos++];
    if (!defined $t) { return 0; }
    if ($t eq '(') {
        my $v = parse_expr();
        $pos++ if $pos < @tok && $tok[$pos] eq ')';
        return $v;
    }
    return $t + 0;
}

while (my $line = <STDIN>) {
    $line =~ s/[\r\n]+$//;
    next if $line =~ /^\s*$/;
    @tok = ($line =~ /(\d+|[-+*\/()])/g);
    $pos = 0;
    my $v = parse_expr();
    print "$v\n";
}

File: solutions/fable-1/fable/javascript/calc.js
'use strict';

const input = require('fs').readFileSync(0, 'utf8');

function evaluate(s) {
  let pos = 0;

  function skipSpaces() {
    while (pos < s.length && /\s/.test(s[pos])) pos++;
  }

  function parseNumber() {
    skipSpaces();
    let start = pos;
    while (pos < s.length && s[pos] >= '0' && s[pos] <= '9') pos++;
    return BigInt(s.slice(start, pos));
  }

  function parseFactor() {
    skipSpaces();
    if (s[pos] === '(') {
      pos++;
      const v = parseExpr();
      skipSpaces();
      if (s[pos] === ')') pos++;
      return v;
    }
    return parseNumber();
  }

  function parseTerm() {
    let v = parseFactor();
    for (;;) {
      skipSpaces();
      const c = s[pos];
      if (c === '*') {
        pos++;
        v = v * parseFactor();
      } else if (c === '/') {
        pos++;
        v = v / parseFactor(); // BigInt division truncates toward zero
      } else {
        return v;
      }
    }
  }

  function parseExpr() {
    let v = parseTerm();
    for (;;) {
      skipSpaces();
      const c = s[pos];
      if (c === '+') {
        pos++;
        v = v + parseTerm();
      } else if (c === '-') {
        pos++;
        v = v - parseTerm();
      } else {
        return v;
      }
    }
  }

  return parseExpr();
}

const out = [];
for (const rawLine of input.split('\n')) {
  const line = rawLine.replace(/\r$/, '');
  if (line.trim() === '') continue;
  out.push(evaluate(line).toString());
}
process.stdout.write(out.join('\n') + (out.length ? '\n' : ''));
